File: apps/web/src/shared/api/messages.ts
import { apiRequest, buildApiUrl } from "./client";

export type MessageReplyReference = {
  id: number;
  author_username: string;
  body_text: string | null;
  deleted_at: string | null;
};

export type MessageAttachment = {
  id: string;
  original_filename: string;
  media_type: string | null;
  size_bytes: number;
  comment_text: string | null;
  content_path: string;
  download_path: string;
  is_image: boolean;
};

export type ConversationMessage = {
  id: number;
  conversation_id: string;
  author_user_id: string | null;
  author_username: string;
  sequence_number: number;
  body_text: string | null;
  reply_to_message_id: number | null;
  reply_to_message: MessageReplyReference | null;
  created_at: string;
  edited_at: string | null;
  deleted_at: string | null;
  is_edited: boolean;
  is_deleted: boolean;
  can_edit: boolean;
  can_delete: boolean;
  attachments: MessageAttachment[];
};

export type ConversationMessageListResponse = {
  conversation_id: string;
  sequence_head: number;
  oldest_loaded_sequence: number | null;
  newest_loaded_sequence: number | null;
  next_before_sequence: number | null;
  has_older: boolean;
  messages: ConversationMessage[];
};

export type ConversationReadResponse = {
  conversation_id: string;
  last_read_sequence_number: number;
  unread_count: number;
};

export type CreateMessagePayload = {
  body_text: string;
  reply_to_message_id?: number;
};

export type CreateAttachmentMessagePayload = {
  body_text?: string;
  reply_to_message_id?: number;
  attachment_comment?: string;
  files: File[];
};

export type EditMessagePayload = {
  body_text: string;
};

export function getAttachmentAssetUrl(path: string) {
  return buildApiUrl(path);
}

export const messagesApi = {
  list(conversationId: string, limit = 50, beforeSequence?: number) {
    return apiRequest<ConversationMessageListResponse>(
      `/api/conversations/${conversationId}/messages`,
      {
        query: { limit, before_sequence: beforeSequence },
      },
    );
  },
  create(conversationId: string, payload: CreateMessagePayload) {
    return apiRequest<ConversationMessage>(
      `/api/conversations/${conversationId}/messages`,
      {
        method: "POST",
        body: JSON.stringify(payload),
      },
    );
  },
  createWithAttachments(conversationId: string, payload: CreateAttachmentMessagePayload) {
    const formData = new FormData();

    if (payload.body_text !== undefined) {
      formData.append("body_text", payload.body_text);
    }

    if (payload.reply_to_message_id !== undefined) {
      formData.append("reply_to_message_id", String(payload.reply_to_message_id));
    }

    if (payload.attachment_comment !== undefined) {
      formData.append("attachment_comment", payload.attachment_comment);
    }

    for (const file of payload.files) {
      formData.append("files", file);
    }

    return apiRequest<ConversationMessage>(
      `/api/conversations/${conversationId}/messages/attachments`,
      {
        method: "POST",
        body: formData,
      },
    );
  },
  edit(messageId: number, payload: EditMessagePayload) {
    return apiRequest<ConversationMessage>(`/api/messages/${messageId}`, {
      method: "PATCH",
      body: JSON.stringify(payload),
    });
  },
  delete(messageId: number) {
    return apiRequest<ConversationMessage>(`/api/messages/${messageId}`, {
      method: "DELETE",
    });
  },
  markRead(conversationId: string) {
    return apiRequest<ConversationReadResponse>(`/api/conversations/${conversationId}/read`, {
      method: "POST",
    });
  },
};

File: apps/api/app/api/schemas/messages.py
from __future__ import annotations

from datetime import datetime
from typing import Annotated
from uuid import UUID

from pydantic import BaseModel, ConfigDict, StringConstraints

MessageBody = Annotated[
    str,
    StringConstraints(
        min_length=1,
        max_length=3072,
    ),
]


class MessageReplyReferenceResponse(BaseModel):
    id: int
    author_username: str
    body_text: str | None
    deleted_at: datetime | None


class MessageAttachmentResponse(BaseModel):
    id: UUID
    original_filename: str
    media_type: str | None
    size_bytes: int
    comment_text: str | None
    content_path: str
    download_path: str
    is_image: bool


class ConversationMessageResponse(BaseModel):
    id: int
    conversation_id: UUID
    author_user_id: UUID | None
    author_username: str
    sequence_number: int
    body_text: str | None
    reply_to_message_id: int | None
    reply_to_message: MessageReplyReferenceResponse | None
    created_at: datetime
    edited_at: datetime | None
    deleted_at: datetime | None
    is_edited: bool
    is_deleted: bool
    can_edit: bool
    can_delete: bool
    attachments: list[MessageAttachmentResponse]


class ConversationMessageListResponse(BaseModel):
    conversation_id: UUID
    sequence_head: int
    oldest_loaded_sequence: int | None
    newest_loaded_sequence: int | None
    next_before_sequence: int | None
    has_older: bool
    messages: list[ConversationMessageResponse]


class ConversationReadResponse(BaseModel):
    conversation_id: UUID
    last_read_sequence_number: int
    unread_count: int


class CreateMessageRequest(BaseModel):
    model_config = ConfigDict(
        json_schema_extra={
            "example": {
                "body_text": "Hello team,\nthis is the first real persisted message.",
                "reply_to_message_id": 12,
            }
        }
    )

    body_text: MessageBody
    reply_to_message_id: int | None = None


class EditMessageRequest(BaseModel):
    model_config = ConfigDict(
        json_schema_extra={
            "example": {
                "body_text": "Edited message body with clarified details.",
            }
        }
    )

    body_text: MessageBody
